plugins/modules/storage/zfs/zfs.py

```diff
@@ -194,19 +194,25 @@ def set_property(self, prop, value):
             self.module.fail_json(msg=err)
 
     def set_properties_if_changed(self):
+        diff = {'before': {}, 'after': {}}
         current_properties = self.get_current_properties()
         for prop, value in self.properties.items():
-            if current_properties.get(prop, None) != value:
+            current_value = current_properties.get(prop, None)
+            if current_value != value:
                 self.set_property(prop, value)
+                diff['before'][prop] = current_value
+                diff['after'][prop] = value
         if self.module.check_mode:
-            return
+            return diff
         updated_properties = self.get_current_properties()
-        for prop in self.properties:
+        for prop in diff['before']:
```

Comment:
I think using `diff['before']` instead of `self.properties` makes the code a bit harder to follow. `self.properties` is the "source of truth" here, so to speak, so I'd suggest keeping this line as it was.

Reply:
I fully understand your concerns. I tried to weight code readability against code optimization and decided for the latter. But I don't mind deciding the other way. Therefore I reverted back this code line and added an if-statement in order to correctly maintain the diff['after'] dictionary. If you're fine with it, please resolve the conversation.

Reply:
I don't see any resolve button; I think it's gone since GitHub considers this comment to be on an "outdated" diff. Regardless, this looks good to me!

Reply:
The GitHub UI unforunately only allows the person who created the PR and persons with sufficient access rights to the repository to resolve comments, but not the person who created them (except if that person falls into one of the former categories). Why is pretty unfortunate IMO :) Anyway, I'll resolve this.

Reply:
@sam-lunt if "Regardless, this looks good to me!" applies to the whole PR, you should add that as a review so it's clear that you approve of it :)

Reply:
Good point, that was ambiguous. I left a top-level comment to that effect ; )